File: internal/actions/runner.go
package actions

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"
	"net/http"
	"strings"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/openforms/openforms/internal/auth"
	"github.com/openforms/openforms/internal/config"
	"github.com/openforms/openforms/internal/definition"
	"github.com/openforms/openforms/internal/definitions"
	"github.com/openforms/openforms/internal/jobs"
	"github.com/openforms/openforms/internal/submissions"
)

type Deps struct {
	Config     config.Config
	Subs       *submissions.Service
	Defs       *definitions.Store
	Auth       *auth.Service
	Mailer     Mailer
	HTTPClient *http.Client
	Pool       *pgxpool.Pool
}

type runner struct{ d Deps }

// Register installs the action job handlers and the failure hook on q.
func Register(q *jobs.Queue, d Deps) {
	if d.HTTPClient == nil {
		d.HTTPClient = &http.Client{Timeout: 10 * time.Second}
	}
	if d.Mailer == nil {
		d.Mailer = NewMailer(d.Config)
	}
	r := &runner{d: d}
	q.Register(KindWebhook, r.webhook)
	q.Register(KindEmail, r.email)
	q.Register(KindAssign, r.assign)
	q.SetFailedHook(r.onFailed)
}

// actionContext is everything a handler needs about the job's submission.
type actionContext struct {
	job        jobs.Job
	payload    Payload
	sub        submissions.Submission
	form       definition.Form
	transition *definition.Transition
	fromState  string
	vars       map[string]any
}

func decodePayload(job jobs.Job) (Payload, error) {
	var p Payload
	if err := json.Unmarshal(job.Payload, &p); err != nil {
		return p, jobs.Permanent(fmt.Errorf("decode action payload: %w", err))
	}
	return p, nil
}

func (r *runner) load(ctx context.Context, job jobs.Job) (*actionContext, error) {
	p, err := decodePayload(job)
	if err != nil {
		return nil, err
	}
	sub, err := r.d.Subs.Get(ctx, job.OrgID, p.SubmissionID)
	if errors.Is(err, submissions.ErrNotFound) {
		return nil, jobs.Permanent(err)
	}
	if err != nil {
		return nil, err
	}
	formRec, err := r.d.Defs.GetFormVersion(ctx, sub.FormVersionID)
	if err != nil {
		return nil, err
	}
	ac := &actionContext{job: job, payload: p, sub: sub, form: formRec.Definition}
	if p.Trigger != TriggerSubmit && sub.WorkflowVersionID != nil {
		wfRec, err := r.d.Defs.GetWorkflowVersion(ctx, *sub.WorkflowVersionID)
		if err != nil {
			return nil, err
		}
		if t, ok := wfRec.Definition.Transition(p.Trigger); ok {
			ac.transition = &t
		}
		if p.EventID != 0 {
			var from *string
			err := r.d.Pool.QueryRow(ctx, `SELECT from_state FROM submission_events WHERE id = $1`, p.EventID).Scan(&from)
			if err != nil && !errors.Is(err, pgx.ErrNoRows) {
				return nil, err
			}
			if from != nil {
				ac.fromState = *from
			}
		}
	}
	ac.vars = TemplateVars(r.d.Config, ac.form, sub, ac.transition)
	return ac, nil
}

func actionMap(a definition.Action) map[string]any {
	b, _ := json.Marshal(a)
	m := map[string]any{}
	_ = json.Unmarshal(b, &m)
	return m
}

func (r *runner) recordSuccess(ctx context.Context, ac *actionContext) error {
	_, err := submissions.InsertEvent(ctx, r.d.Pool, submissions.Event{
		SubmissionID: ac.sub.ID,
		OrgID:        ac.sub.OrgID,
		Type:         "action_succeeded",
		ActorType:    "system",
		ActorName:    SystemActorName,
		Payload:      map[string]any{"action": actionMap(ac.payload.Action), "trigger": ac.payload.Trigger, "jobId": ac.job.ID},
	})
	return err
}

// onFailed writes an action_failed event when an action job ends failed.
func (r *runner) onFailed(ctx context.Context, job jobs.Job, runErr error) {
	if !strings.HasPrefix(job.Kind, "action.") {
		return
	}
	p, err := decodePayload(job)
	if err != nil {
		slog.ErrorContext(ctx, "actions: failed job has undecodable payload", "job", job.ID, "err", err)
		return
	}
	_, err = submissions.InsertEvent(ctx, r.d.Pool, submissions.Event{
		SubmissionID: p.SubmissionID,
		OrgID:        job.OrgID,
		Type:         "action_failed",
		ActorType:    "system",
		ActorName:    SystemActorName,
		Payload:      map[string]any{"error": runErr.Error(), "action": actionMap(p.Action), "trigger": p.Trigger, "jobId": job.ID},
	})
	if err != nil {
		slog.ErrorContext(ctx, "actions: record action_failed", "job", job.ID, "err", err)
	}
}

File: src/openforms/server/actions/runner.py
"""Action job handlers and the success/failure events they record."""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from typing import Any

import httpx
import sqlalchemy as sa

from ...definition import Form, Transition
from ...settings import Settings
from ..database.engine import Database
from ..exceptions import NotFound
from ..models import definitions as defs
from ..models import submissions as subs
from ..services.jobs import Job, Permanent, Queue
from .mailer import Mailer, new_mailer
from .payload import KIND_ASSIGN, KIND_EMAIL, KIND_WEBHOOK, SYSTEM_ACTOR_NAME, TRIGGER_SUBMIT, Payload

log = logging.getLogger("openforms.actions")


@dataclass
class ActionDeps:
    settings: Settings
    db: Database
    mailer: Mailer | None = None
    http_client: httpx.AsyncClient | None = None


@dataclass
class ActionContext:
    job: Job
    payload: Payload
    sub: subs.Submission
    form: Form
    transition: Transition | None = None
    from_state: str = ""
    vars: dict[str, Any] = field(default_factory=dict)


def decode_payload(job: Job) -> Payload:
    try:
        return Payload.from_dict(job.payload)
    except (KeyError, ValueError, TypeError) as e:
        raise Permanent(ValueError(f"decode action payload: {e}")) from None


class Runner:
    def __init__(self, d: ActionDeps) -> None:
        self.d = d
        self.mailer: Mailer = d.mailer or new_mailer(d.settings)
        self.http = d.http_client or httpx.AsyncClient(timeout=10.0)

    async def load(self, job: Job) -> ActionContext:
        from .vars import template_vars

        p = decode_payload(job)
        async with self.d.db.session() as s:
            try:
                sub = await subs.get(s, job.org_id, p.submission_id)
            except NotFound as e:
                raise Permanent(e) from None
            form = (await defs.get_form_version(s, sub.form_version_id)).definition
            ac = ActionContext(job=job, payload=p, sub=sub, form=form)
            if p.trigger != TRIGGER_SUBMIT and sub.workflow_version_id is not None:
                wf = (await defs.get_workflow_version(s, sub.workflow_version_id)).definition
                ac.transition = wf.transition(p.trigger)
                if p.event_id:
                    frm = (
                        await s.execute(
                            sa.text("SELECT from_state FROM submission_events WHERE id = :id"), {"id": p.event_id}
                        )
                    ).scalar()
                    ac.from_state = frm or ""
        ac.vars = template_vars(self.d.settings, ac.form, sub, ac.transition)
        return ac

    async def record_success(self, ac: ActionContext) -> None:
        async with self.d.db.transaction() as s:
            await subs.insert_event(
                s,
                subs.Event(
                    submission_id=ac.sub.id,
                    org_id=ac.sub.org_id,
                    type=subs.EVENT_ACTION_SUCCEEDED,
                    actor_type="system",
                    actor_name=SYSTEM_ACTOR_NAME,
                    payload={"action": ac.payload.action.to_dict(), "trigger": ac.payload.trigger, "jobId": ac.job.id},
                ),
            )

    async def on_failed(self, job: Job, err: BaseException) -> None:
        """Write an ``action_failed`` event when an action job ends failed."""
        if not job.kind.startswith("action."):
            return
        try:
            p = decode_payload(job)
        except Permanent:
            log.error("actions: failed job %s has an undecodable payload", job.id)
            return
        async with self.d.db.transaction() as s:
            await subs.insert_event(
                s,
                subs.Event(
                    submission_id=p.submission_id,
                    org_id=job.org_id,
                    type=subs.EVENT_ACTION_FAILED,
                    actor_type="system",
                    actor_name=SYSTEM_ACTOR_NAME,
                    payload={"error": str(err), "action": p.action.to_dict(), "trigger": p.trigger, "jobId": job.id},
                ),
            )

    async def webhook(self, job: Job) -> None:
        from .webhook import run_webhook

        await run_webhook(self, job)

    async def email(self, job: Job) -> None:
        from .email import run_email

        await run_email(self, job)

    async def assign(self, job: Job) -> None:
        from .assign import run_assign

        await run_assign(self, job)


def register(queue: Queue, d: ActionDeps) -> Runner:
    """Install the action job handlers and the failure hook on ``queue``."""
    r = Runner(d)
    queue.register(KIND_WEBHOOK, r.webhook)
    queue.register(KIND_EMAIL, r.email)
    queue.register(KIND_ASSIGN, r.assign)
    queue.set_failed_hook(r.on_failed)
    return r
